### PR TITLE
MRG: Allow reading and plotting of xfit dipoles

### LINKED ISSUE #2935: Reading Xfit dipole file
@Eric89GXL attaching xfit .fif in compressed file containing single ecd waveform and goodness of fit time series. You can also find [here] (https://www.dropbox.com/s/lurthfo1hy9tzvg/fam_115_LH.fif?dl=0) if attachment doesn't work. 
[fam_115_LH.zip](https://github.com/mne-tools/mne-python/files/143288/fam_115_LH.zip)

Like I said `read_meas_info(fid, tree, clean_bads=True)` seems to get at `meas` but not `info` b/c there is no isotrak, unable to make much sense out of the TAG structure. Although, 't = mne.io.tag.read_tag(fid, pos=3, shape=204)' implies some hope....

> File "/home/ktavabi/Projects/mne-python/mne/io/tag.py", line 501, in read_tag
>     tag.data = _read_matrix(fid, tag, shape, rlims, matrix_coding)
>   File "/home/ktavabi/Projects/mne-python/mne/io/tag.py", line 237, in _read_matrix
>     raise ValueError('Row reading not implemented for matrices '
> ValueError: Row reading not implemented for matrices yet

### PR DESCRIPTION
**Pull request by Eric89GXL**
Closes #2935.

Ready for review/merge from my end.

### CODE REVIEW

**Comment by Eric89GXL** [comment]
Sample output on my sistem of `evoked.plot`:

figure_1

File: mne/io/constants.py [hunk 1]
```diff
@@ -325,6 +327,8 @@ def __setattr__(self, attr, val):
 FIFF.FIFF_PROJ_ITEM_VECTORS      = 3415
 FIFF.FIFF_PROJ_ITEM_DEFINITION   = 3416
 FIFF.FIFF_PROJ_ITEM_CH_NAME_LIST = 3417
+#   XPlotter
+FIFF.FIFF_XPLOTTER_LAYOUT        =3501  # string - "Xplotter layout tag"
```
Comment (agramfort) [inline mne/io/constants.py]: pep8

**Comment by agramfort** [comment]
why not using our Dipole class?

can we make our dipole fits compatible with xfit ?

**Comment by Eric89GXL** [comment]
They are actually quite different. Our dipole class fits a single dipole to each time point. This one I think fits a single dipole location for all time points, and gives the resulting time course. The saving format is also completely differetn -- xfit produces an evoked file, and our `.dip` class has not very similar information actually.

**Comment by Eric89GXL** [comment]
@agramfort comments addressed

**Comment by kambysese** [comment]
@agramfort 
> why not using our Dipole class?

Beside desperation, really a lack of familiarity and not having a pipeline in place for ECD fitting using mne-python on my end.

@Eric89GXL 
This particular fit is essentially a [SSP](http://www.sciencedirect.com/science/article/pii/0013469495000646) approach. Xfit has several 'flavors' of ECD fitting routines including sequential (most likely the mne-python dipole class), fixed, and fixed-orientation fitting options

File: mne/evoked.py [hunk 2]
```diff
@@ -182,6 +181,12 @@ def __init__(self, fname, condition=None, baseline=None, proj=True,
                 elif my_kind == FIFF.FIFF_CH_INFO:
                     tag = read_tag(fid, pos)
                     chs.append(tag.data)
+                elif my_kind == FIFF.FIFF_FIRST_TIME:
+                    tag = read_tag(fid, pos)
+                    first_time = float(tag.data)
+                elif my_kind == FIFF.FIFF_NO_SAMPLES:
+                    tag = read_tag(fid, pos)
+                    nsamp = int(tag.data)
```
Comment (kambysese) [inline mne/evoked.py]: @Eric89GXL out of curiosity is this where the reading magic happens? once the fiff constants and meas_info vars are defined.

Comment (Eric89GXL) [inline mne/evoked.py]: There isn't really any magic. There are four things happening in this PR that we needed:

1. Recognize that the file looks like an evoked file does on disk.
2. Add a couple of missing FIFF constants.
3. Enhance the plotting and pick functions to deal with new constants.
4. Fix the evoked reading code to deal with a different way of defining the start and end points -- we usually use `first` and `last`, here we have `first_time` (with `sfreq`) and `no_samples` to get the same information.

This particular point in the code is where 4) happens.

Comment (kambysese) [inline mne/evoked.py]: How did you figure out (1)?

Comment (Eric89GXL) [inline mne/evoked.py]: I did:
```Bash
$ mne show_fiff ../mne-testing-data/misc/fam_115_LH.fif 
100 = FIFFB_MEAS
    103 = FIFF_BLOCK_ID (20) = {'secs': 1455227835, 'version' ... dict len=4
    109 = FIFF_PARENT_FILE_ID (20) = {'secs': 1455217825, 'version' ... dict len=4
    110 = FIFF_PARENT_BLOCK_ID (20) = {'secs': 1455213477, 'version' ... dict len=4
    101 = FIFFB_MEAS_INFO
        204 = FIFF_MEAS_DATE (8) = [1455227835     561810] ... array size=2
        201 = FIFF_SFREQ (4) = [ 499.99996948]
        200 = FIFF_NCHAN (4) = [2]
        x2: 203 = FIFF_CH_INFO (96) = {'loc': array([ 0.,  0.,  0.,  ... dict len=11
        3501 = ??? (102) = 0 4.000000 0 6.000000¶1 0 3.00 ... str len=102
    103 = FIFFB_PROCESSED_DATA
        104 = FIFFB_EVOKED
            229 = FIFF_FIRST_TIME (4) = [-0.1]
            228 = FIFF_NO_SAMPLES (4) = [500]
            105 = FIFFB_ASPECT
                210 = FIFF_ASPECT_KIND (4) = [200]
                207 = FIFF_NAVE (4) = [1]
                x2: 302 = FIFF_EPOCH (2000) = [  1.73851430e-01   1.07006311 ... array size=500
```
The `FIFFB_EVOKED` tipped me off.

Comment (kambysese) [inline mne/evoked.py]: Oh man, it all makes sense now I completely missed that tag yesterday. Thanks.

**Comment by agramfort** [comment]
I am bit worried to store in an Evoked object something that is not sensor space data and actual evoked data. It's convenient for xfit but misleading from a user stand point. To me it should be exposed as a flavor of our Dipole class with a plot method that corresponds to dipole fit (source loc / time course).

sorry to be a pain...

**Comment by kambysese** [comment]
Fair enough, frankly after this attempt I will use the ecd fitting in mne-python so if this is too much of a pain @Eric89GXL I can just temporarily work of this branch. Alternatively if this additional functionality is of any use to anyone in the future I suppose you can hook the reading via `evoked.py` in this pr into `read_dipole`. Just a thought. LMK, in case [here](https://www.dropbox.com/s/c7utgz7hys1vscl/All_kubi_td115_RH.dip?dl=0) is a corresponding dip file.

**Comment by Eric89GXL** [comment]
> To me it should be exposed as a flavor of our Dipole class with a plot method that corresponds to dipole fit (source loc / time course).

I see your point, but making this a part of the `Dipole` class has several disadvantages. To realize this, just think about the difference between a typical `Evoked` class -- which this operates seamlessly as under the hood -- and the current `Dipole` class. At a minimum, there is:

1. Mismatch between saving and loading functions. Right now we have `read_evokeds` and `write_evokeds`, and they work with this new `.fif` file with minimal changes. `.dip` files are text files, and can't easily be expanded.

2. Mismatch in plotting functions. All the Evoked plotting functions should work (or can be expanded to work) with this. I don't think you can say the same of the `Dipole` class as it is. Making this part of the `Dipole` class means some of the `dip.plot_*` would only work for `.dip`-type dipoles and others for xfit-style `.fif` dipoles.

3. Mismatch in data represented. `Dipole`/`.dip` has some data that `.fif` does not have, and vice-versa.

So I don't think `Dipole` absorbing this class makes things simpler from a user or API point of view.

Maybe instead of making it use `read_evokeds` and `write_evokeds` and the class `Evoked`, I can make a `read_xfit_dipole` (or so) and matching write function, and class `XfitDipole` or so. This could wrap almost directly to `read_evokeds` and `write_evokeds` and `Evoked` with minimal changes, I think. Or maybe read/write_dipole should triage under the hood, but return a `Dipole` or `XfitDipole` instance, based on the input type.

Maybe some day we will be able to have a single `Dipole` class that represents both MNE-style `.dip` files and xfit `.fif` files, but right I find it confusing to think about how they'd function properly in the same class.

**Comment by agramfort** [comment]
>
> Maybe instead of making it use read_evokeds and write_evokeds and the
> class Evoked, I can make a read_xfit_dipole (or so) and matching write
> function, and class XfitDipole or so. This could wrap almost directly to
> read_evokeds and write_evokeds and Evoked with minimal changes, I think.
> Or maybe read/write_dipole should triage under the hood, but return a
> Dipole or XfitDipole instance, based on the input type.
>
I was thinking the same. +1 for a read_xfit_dipole and a new XfitDipole
class (in dipole.py).

Ideally the plot method of XfitDipole should be specific to the fact that
it corresponds to source information. Any option to show it in source space?

anyone prefers a single smarter read_dipole function?

**Comment by Eric89GXL** [comment]
@kambysese you mentioned that XFit has multiple output modes -- could you do a dipole fit (preferably for the same data -- and if so can you also upload the anonymized -ave.fif?) using each mode and upload the FIF files somewhere? That would help make sure we deal with all outputs. We might as well try to get it right.

**Comment by Eric89GXL** [comment]
From talking to @kambysese, it sounds like only `.dip` and `.fif` files of the type added here are going to be useful. I've made a new `DipoleXfit` class to deal with the FIF case.

I ended up going with augmenting the `read_dipole` function because it was pretty straightforward.

Ready for review/merge from my end.

File: mne/dipole.py [hunk 3]
```diff
@@ -214,6 +227,70 @@ def __len__(self):
...
+        ----------
+        fname : str
+            The name of the .fif file.
+        verbose : bool, str, int, or None
+            If not None, override default verbose level (see mne.verbose).
+        """
+        _check_fname(fname, overwrite=True, must_exist=False)
+        if not fname.endswith('.fif') and not fname.endswith('.fif.gz'):
+            raise ValueError('filename must end with .fif or .fif.gz')
+        _write_evokeds(fname, self, check=False)
```
Comment (agramfort) [inline mne/dipole.py]: Should promote files that end with -dip.fif ?

Comment (Eric89GXL) [inline mne/dipole.py]: I don't think XFit enforces anything, but I suppose we could add it

Comment (Eric89GXL) [inline mne/dipole.py]: (probably only check on write, not on read)

**Comment by Eric89GXL** [comment]
@agramfort check added

File: mne/dipole.py [hunk 4]
```diff
@@ -214,6 +227,70 @@ def __len__(self):
...
+        return self.info['ch_names']
+
+    @verbose
+    def save(self, fname, verbose=None):
+        """Save dipole in a .fif file
+
+        Parameters
+        ----------
+        fname : str
+            The name of the .fif file.
```
Comment (agramfort) [inline mne/dipole.py]: "It is recommended to end the file with -dip.fif to make it explicit that the file contains dipole information."

File: mne/dipole.py [hunk 5]
```diff
@@ -214,6 +227,70 @@ def __len__(self):
...
+        verbose : bool, str, int, or None
+            If not None, override default verbose level (see mne.verbose).
+        """
+        check_fname(fname, 'DipoleFixed', ('-dip.fif', '-dip.fif.gz'))
+        if not fname.endswith('.fif') and not fname.endswith('.fif.gz'):
+            raise ValueError('filename must end with .fif or .fif.gz')
```
Comment (agramfort) [inline mne/dipole.py]: to me the Exception should be in check_fname function.

File: mne/dipole.py [hunk 6]
```diff
@@ -214,6 +227,70 @@ def __len__(self):
...
+        _write_evokeds(fname, self, check=False)
+
+    def plot(self, show=True):
+        """Plot dipole data
+
+        Parameters
+        ----------
+        show : bool
+            Call pyplot.show() at the end or not.
```
Comment (agramfort) [inline mne/dipole.py]: Returns

**Comment by agramfort** [comment]
that's it for me.

thanks @Eric89GXL

**Comment by Eric89GXL** [comment]
@agramfort comments addressed and CIs look like they'll be happy. +1 for merge from you?

**Comment by agramfort** [comment]
thanks @Eric89GXL

**Pull request merged**